services/121-service/src/payments/fsp-integration/safaricom/safaricom.service.ts

```diff
@@ -173,104 +111,77 @@ export class SafaricomService
-    } else {
-      paTransactionResult.status = StatusEnum.error;
-      payload.status = StatusEnum.error;
-      paTransactionResult.message = result.errorMessage;
+    if (result && result.ResponseCode !== '0') {
+      throw new Error(result.errorMessage);
     }
 
-    paTransactionResult.customData = {
-      requestResult: result,
+    return {
+      amountTransferredInMajorUnit: payload.Amount,
+      customData: {
+        requestResult: result,
+      },
     };
-    return paTransactionResult;
   }
 
-  public async processSafaricomRequest(
-    payload,
-    paymentRequestResultPerPa,
+  public async createAndSaveSafaricomTransferData(
+    safaricomDoTransferResult: DoTransferReturnParams,
     transaction: TransactionEntity,
   ): Promise<any> {
-    const payloadResult = Object.fromEntries(
-      Object.entries(payload).map(([k, v]) => [
-        k.charAt(0).toLowerCase() + k.slice(1),
-        v,
-      ]),
-    );
+    const safaricomTransferEntity = new SafaricomTransferEntity();
 
-    const safaricomCustomData = { ...paymentRequestResultPerPa.customData };
-
-    payloadResult.requestResult = safaricomCustomData.requestResult;
-    payloadResult.conversationID =
+    const safaricomCustomData = { ...safaricomDoTransferResult.customData };
+    safaricomTransferEntity.mpesaConversationId =
       safaricomCustomData &&
       safaricomCustomData.requestResult &&
-      safaricomCustomData.requestResult.ConversationID
-        ? safaricomCustomData.requestResult.ConversationID
+      safaricomCustomData.requestResult['ConversationID']
+        ? safaricomCustomData.requestResult['ConversationID']
         : 'Invalid Request';
-    payloadResult.transaction = transaction;
-    await this.safaricomRequestRepository.save(payloadResult);
+    safaricomTransferEntity.originatorConversationId =
+      safaricomCustomData &&
+      safaricomCustomData.requestResult &&
+      safaricomCustomData.requestResult['OriginatorConversationID']
+        ? safaricomCustomData.requestResult['OriginatorConversationID']
+        : 'Invalid Request';
+
+    safaricomTransferEntity.transactionId = transaction.id;
+
+    await this.safaricomTransferRepository.save(safaricomTransferEntity);
   }
 
-  public async processSafaricomResult(
-    safaricomPaymentResultData: any,
-    attempt = 1,
+  public async processSafaricomCallback(
+    safaricomPaymentResultData: SafaricomTransferCallbackJobDto,
+    _attempt = 1,
   ): Promise<void> {
-    const safaricomDbRequest = await this.safaricomRequestRepository
-      .createQueryBuilder('safaricom_request')
-      .leftJoinAndSelect('safaricom_request.transaction', 'transaction')
+    const job = await this.safaricomCallbackQueue.add(
+      ProcessNamePayment.callbackPayment,
+      safaricomPaymentResultData,
+    );
+    await this.redisClient.sadd(getRedisSetName(job.data.programId), job.id);
+  }
+
+  public async getSafaricomTransferByOriginatorConversationId(
```

Comment:
This function belongs in a (custom) repo.

Reply:
processed